pkg/platform/api/buildplanner/model/buildplan.go

```diff
@@ -65,30 +70,134 @@ func (o Operation) String() string {
 }
 
 // BuildPlan is the top level object returned by the build planner. It contains
-// the project, commit, and build.
-type BuildPlan struct {
+// the commit and build.
+type BuildPlan interface {
+	Build() (*Build, error)
+	CommitID() (strfmt.UUID, error)
+}
+
+func NewBuildPlanResponse(owner, project string) BuildPlan {
+	if owner != "" && project != "" {
+		return &BuildPlanByProject{}
+	}
+	return &BuildPlanByCommit{}
+}
+
+type BuildPlanByProject struct {
 	Project *Project `json:"project"`
+	*Error
+}
+
+func (b *BuildPlanByProject) Build() (*Build, error) {
+	if b.Project == nil {
+		return nil, errs.New("Project is nil")
+	}
+
+	if b.Project.Error != nil && b.Project.Message != "" {
```

Comment:
This feels wrong; if Error isn't nil but Message is empty then we disregard the error? Same for the Commit error check further down. If this is intended please add a comment explaining.

Reply:
The error is just a message. It's an embedded struct in the buildplanner types. If the error is nil then the message should also be.

I've updated how we access the message value.

Reply:
Ok, but then Error being non-nil and message being empty IS an error condition which the code as written doesn't seem to address.